internal/pkg/csumio/csumwriter_test.go

```diff
@@ -97,3 +97,233 @@ func TestWriter_Success(t *testing.T) {
+			preLens:     nil,
+			targetLen:   pEqual,
+			expectedChunkLens: []int{
+				chunkSize,
+			},
+		},
+		{
+			name:        "case06_pEqual_capacityLess_offsetNonZero",
+			description: "p == chunk, capacity < p, offset non-zero",
+			preLens: []int{
+				chunkSize,
+				almostFull,
+			},
+			targetLen: pEqual,
+			expectedChunkLens: []int{
+				chunkSize,
+				chunkSize,
+				almostFull,
+			},
+		},
+		{
+			name:        "case07_pSmaller_capacityGreater_offsetZero",
+			description: "p < chunk, capacity > p, offset zero",
+			preLens:     nil,
+			targetLen:   pSmaller,
+			expectedChunkLens: []int{
+				pSmaller,
+			},
+		},
+		{
+			name:        "case08_pSmaller_capacityEqual_offsetNonZero",
+			description: "p < chunk, capacity == p, offset non-zero",
+			preLens: []int{
+				chunkSize,
+				singleByte,
+			},
+			targetLen: pSmaller,
+			expectedChunkLens: []int{
+				chunkSize,
+				chunkSize,
+			},
+		},
+		{
+			name:        "case09_pSmaller_capacityLess_offsetZero",
+			description: "p < chunk, capacity < p, offset zero",
+			preLens: []int{
+				almostFull,
+			},
+			targetLen: pSmaller,
+			expectedChunkLens: []int{
+				chunkSize,
+				twoBytesShort,
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			// Arrange
+			dataRecorder := &chunkRecorder{}
+			checksumBuffer := &bytes.Buffer{}
+			cw, err := csumio.NewWriter(dataRecorder, checksumBuffer, chunkSize)
+			require.NoError(t, err)
+
+			currentDataByte := 0
+			expectedData := make([]byte, 0)
+
+			write := func(length int) {
+				data := bytes.Repeat([]byte{byte(currentDataByte)}, length)
+				currentDataByte++
+				n, err := cw.Write(data)
+				require.NoError(t, err)
+				require.Equal(t, length, n)
+				expectedData = append(expectedData, data...)
+			}
+
+			for _, preLen := range tc.preLens {
+				write(preLen)
+			}
+
+			// Act
+			write(tc.targetLen)
+			err = cw.Close()
+			require.NoError(t, err)
+
+			// Assert
+			t.Log(tc.description)
+			require.Equal(t, expectedData, dataRecorder.bytes())
+			assert.Equal(t, tc.expectedChunkLens, dataRecorder.lengths())
+			assertChecksumsMatch(t, dataRecorder.chunks(), checksumBuffer.Bytes())
+		})
+	}
+}
+
+type chunkRecorder struct {
+	writes [][]byte
+}
+
+func (cr *chunkRecorder) Write(p []byte) (int, error) {
+	chunk := make([]byte, len(p))
+	copy(chunk, p)
+	cr.writes = append(cr.writes, chunk)
+	return len(p), nil
+}
+
+func (cr *chunkRecorder) bytes() []byte {
+	total := 0
+	for _, w := range cr.writes {
+		total += len(w)
+	}
+	merged := make([]byte, 0, total)
+	for _, w := range cr.writes {
+		merged = append(merged, w...)
+	}
+	return merged
+}
+
+func (cr *chunkRecorder) lengths() []int {
+	lens := make([]int, len(cr.writes))
+	for i, w := range cr.writes {
+		lens[i] = len(w)
+	}
+	return lens
+}
+
+func (cr *chunkRecorder) chunks() [][]byte {
+	out := make([][]byte, len(cr.writes))
+	copy(out, cr.writes)
+	return out
+}
```

Comment:
I don't think we have to copy the slice. How about just returning `cr.writes` directly?